pyramid_openapi3/__init__.py

```diff
@@ -292,41 +281,54 @@ def register() -> None:
         if hupper.is_active():  # pragma: no cover
             hupper.get_reloader().watch_files(list(path.parent.iterdir()))
 
-        spec_dict = read_yaml_file(path)
+        spec_dict, _ = read_from_filename(str(path))
         spec_url = path.as_uri()
         validate_spec(spec_dict, spec_url=spec_url)
-        spec = create_spec(spec_dict, spec_url=spec_url)
+        spec = Spec.create(spec_dict, url=spec_url)
 
         config.add_static_view(route, str(path.parent), permission=permission)
         config.add_route(route_name, f"{route}/{path.name}")
 
-        custom_formatters = config.registry.settings.get("pyramid_openapi3_formatters")
-        custom_deserializers = config.registry.settings.get(
-            "pyramid_openapi3_deserializers"
+        config.registry.settings[apiname] = _create_api_settings(
+            config, filepath, route_name, spec
         )
-
-        config.registry.settings[apiname] = {
-            "filepath": filepath,
-            "spec_route_name": route_name,
-            "spec": spec,
-            "request_validator": RequestValidator(
-                spec,
-                custom_formatters=custom_formatters,
-                custom_media_type_deserializers=custom_deserializers,
-            ),
-            "response_validator": ResponseValidator(
-                spec,
-                custom_formatters=custom_formatters,
-                custom_media_type_deserializers=custom_deserializers,
-            ),
-        }
         config.registry.settings.setdefault("pyramid_openapi3_apinames", []).append(
             apiname
         )
 
     config.action((f"{apiname}_spec",), register, order=PHASE0_CONFIG)
 
 
+def _create_api_settings(
+    config: Configurator, filepath: str, route_name: str, spec: Spec
+) -> t.Dict:
+    custom_formatters = config.registry.settings.get("pyramid_openapi3_formatters")
+    custom_deserializers = config.registry.settings.get(
+        "pyramid_openapi3_deserializers"
+    )
+
+    media_type_deserializers_factory = MediaTypeDeserializersFactory(
+        custom_deserializers=custom_deserializers
+    )
+    schema_unmarshallers_factory = SchemaUnmarshallersFactory(
+        OAS31Validator, custom_formatters=custom_formatters
```

Comment:
@zupo After doing some digging it appears that Swagger UI does not have full OpenApi 3.1 support yet (`openapi-core` does)

<img width="672" alt="Screenshot 2022-12-05 at 16 40 21" src="https://user-images.githubusercontent.com/35456894/205664345-d62fffb0-a2ee-448a-ae83-7bee20839017.png">

Github Issue: https://github.com/swagger-api/swagger-ui/issues/5891

We will likely have to remain on the `OAS30Validator` validator until support is added to Swagger UI

Reply:
I have reverted back to OpenApi 3.0.x in this PR (What we support currently and default behaviour for `openapi-core`).

I'll create a separate issue on this repo for OpenApi 3.1 support linking it to the Swagger UI issue. That way if anyone else wants something to track for progress then it is here